### PR TITLE
CUDA 11.4 fix for changed API (cudnnGetConvolutionForward/BackwardAlgorithm)

### PR DESCRIPTION
**Pull request by sigfrid696**
I tested the fix and it is working in c++ inference env (no more CUDNN_STATUS_NOT_SUPPORTED error)

### CODE REVIEW

**Comment by nietras** [comment]
@sigfrid696 thanks! 👍 I hope to look at this, build it, test it and release new version next week.

Linking to issue that describes the details:
https://github.com/microsoft/CNTK/issues/3835

File: Source/Math/CuDnnConvolutionEngine.cu [hunk 1]
```diff
@@ -286,19 +286,34 @@ protected:
                 return cudnnGetConvolutionForwardAlgorithm(*m_cudnn, m_inT, *m_kernelT, *m_conv, m_outT, CUDNN_CONVOLUTION_FWD_SPECIFY_WORKSPACE_LIMIT, workspace.BufferSize(), &algo);
             return cudnnGetConvolutionForwardAlgorithm(*m_cudnn, m_inT, *m_kernelT, *m_conv, m_outT, CUDNN_CONVOLUTION_FWD_NO_WORKSPACE, 0, &algo);*/
 
-			// 2020.12.09 - mj.jo
-			// cuda 11.1
+			// 2021.08.30 - sigfrid696
+            // cuda 11.4
             int res_count = 0;
-            cudnnConvolutionFwdAlgoPerf_t fwd_perf;
```
Comment (nietras) [inline Source/Math/CuDnnConvolutionEngine.cu:292]: @sigfrid696 would you mind putting 100 in a local const value?

File: Source/Math/CuDnnConvolutionEngine.cu [hunk 2]
```diff
@@ -595,7 +636,7 @@ private:
             if (m_forceDeterministicAlgorithms)
             {
                 workspace.Resize((algo.DeterministicAlgoWorkspaceSize + sizeof(ElemType) - 1) / sizeof(ElemType), 1, 0, false);
-                CUDNN_CALL(deterministicFinder(calgo, algoPerf));
+				CUDNN_CALL(deterministicFinder(calgo, algoPerf));
```
Comment (nietras) [inline Source/Math/CuDnnConvolutionEngine.cu]: @sigfrid696 undo whitespace edits here and beloq

File: Source/Math/CuDnnConvolutionEngine.cu (hunk 1 above)
Comment (nietras) [reply Source/Math/CuDnnConvolutionEngine.cu:292]: @sigfrid696 there already exists a constant `MaxAlgoCount`, I am doing these changes myself now and will probably file a PR to your PR. Hope you don't mind. :)

**Comment by sigfrid696** [comment]
@nietras thank you!
In respect to some of your changes:
- I restored back a value of 100 in cudnnGetConvolution...v7 functions, because with MaxAlgoCount value (10) we're not guaranteed to obtain an algo, among the other returned by the function, with needed workspace size 0. I tested it in my application. On the other hand I don't want to increase MaxAlgoCount, because I don't want to change behavior of other parts (not changed by the mods)
- I restored double err/err0 return, because an error is returned also if the loop cannot find an algo with needed ws 0. The same pattern can be found in other part of the module (not changed by the mods)

**Comment by nietras** [comment]
@sigfrid696 thanks! Clearly I haven't fully understood your changes it seems, so I have some questions.

With respect to algo count, I checked and there should not be more than 10 available algos, this can be seen in the enum definitions. But given your tests showed this I wonder how that could be?

The latter I will have to check again, but if it's the same pattern used elsewhere we can just leave it.

**Comment by nietras** [comment]
I see my mistake with regards to 'err0' now. My bad. I'd still like a local const for 100 though, but I can do that after.

**Comment by nietras** [comment]
A minor thing is you are using tabs per your editor but code uses spaces, which is also why alignment doesn't match here on GH. 😅

**Review by nietras — APPROVED** [review]
minor nits can be addressed after

File: Source/Math/CuDnnConvolutionEngine.cu (hunk 1 above)
Comment (sigfrid696) [reply Source/Math/CuDnnConvolutionEngine.cu:292]: No problem :)

Comment (sigfrid696) [reply Source/Math/CuDnnConvolutionEngine.cu:292]: > No problem :) 

If you want to open a pull request with these mods I'll report back on my repo. Sorry out of home now and had no time to do the mods

**Comment by sigfrid696** [comment]
> A minor thing is you are using tabs per your editor but code uses spaces, which is also why alignment doesn't match here on GH. 😅

I see... I remember that I didn't change the c Lang formatter already defined in cntk. Probably it was a previous setting in my Vs studio installation

**Comment by sigfrid696** [comment]
> @sigfrid696 thanks! Clearly I haven't fully understood your changes it seems, so I have some questions.
> 
> With respect to algo count, I checked and there should not be more than 10 available algos, this can be seen in the enum definitions. But given your tests showed this I wonder how that could be?
> 
> The latter I will have to check again, but if it's the same pattern used elsewhere we can just leave it. 

I tried to respond on the other thread :)

**Pull request merged**